### PR TITLE
feat(state-sync): Pass sync adapter to client

### PR DESCRIPTION
**Pull request by VanBarbascu**
Start one actor per shard. 
Send start message from client.

### CODE REVIEW

**Comment by VanBarbascu** [comment]
`run_sync_step` is called excessively. [Gragfana](https://nearinc.grafana.net/goto/HN4AFB7SR?orgId=1)

**Comment by VanBarbascu** [comment]
issue [#9625](https://github.com/near/nearcore/issues/9625)

File: chain/client/src/sync/adapter.rs [hunk 1]
```diff
@@ -1,7 +1,8 @@
 use super::sync_actor::SyncActor;
-use actix::{dev::ToEnvelope, Actor, MailboxError, Message};
+use actix::{dev::ToEnvelope, prelude::SendError, Actor, Message};
```
Comment (nikurt) [inline chain/client/src/sync/adapter.rs:2]: https://github.com/near/nearcore/blob/master/docs/practices/style.md#import-granularity

File: chain/client/src/test_utils/setup.rs [hunk 2]
```diff
@@ -162,6 +162,11 @@ pub fn setup(
         chain_genesis,
         epoch_manager,
         shard_tracker,
+        Arc::new(RwLock::new(SyncAdapter::new(
```
Comment (nikurt) [inline chain/client/src/test_utils/setup.rs:165]: Extract the expression into a variable. As it is now, it breaks the flow of reading arguments of a function.

File: chain/client/src/test_utils/setup.rs [hunk 3]
```diff
@@ -929,10 +934,15 @@ pub fn setup_client_with_runtime(
     );
     config.epoch_length = chain_genesis.epoch_length;
     let mut client = Client::new(
-        config,
+        config.clone(),
         chain_genesis,
         epoch_manager,
         shard_tracker,
+        Arc::new(RwLock::new(SyncAdapter::new(
```
Comment (nikurt) [inline chain/client/src/test_utils/setup.rs:941]: Extract into a variable.

File: chain/client/src/client.rs [hunk 4]
```diff
@@ -2711,3 +2751,20 @@ impl Client {
         Ok(ret)
     }
 }
+
+impl Drop for Client {
+    fn drop(&mut self) {
+        // Stop the running actors.
```
Comment (nikurt) [inline chain/client/src/client.rs:2757]: What happens if you don't?

File: chain/client/src/client.rs [hunk 5]
```diff
@@ -2711,3 +2751,20 @@ impl Client {
...
+        let epoch_id = self.chain.store().head().expect("Cannot get chain head.").epoch_id;
+        let shard_layout =
+            self.epoch_manager.get_shard_layout(&epoch_id).expect("Cannot get shard layout.");
```
Comment (nikurt) [inline chain/client/src/client.rs:2760]: Using `epoch_manager` during `drop()` seems complicated.
How about asking `state_sync_adapter` to drop all handles it owns?

File: chain/client/src/client.rs [hunk 6]
```diff
@@ -2382,14 +2403,33 @@ impl Client {
...
-                        BlocksCatchUpState::new(sync_hash, epoch_id),
+                        BlocksCatchUpState::new(sync_hash, epoch_id.clone()),
                     )
                 });
 
             // For colour decorators to work, they need to printed directly. Otherwise the decorators get escaped, garble output and don't add colours.
             debug!(target: "catchup", ?me, ?sync_hash, progress_per_shard = ?format_shard_sync_phase_per_shard(&shards_to_split, false), "Catchup");
             let use_colour = matches!(self.config.log_summary_style, LogSummaryStyle::Colored);
 
+            let tracking_shards: Vec<u64> =
```
Comment (nikurt) [inline chain/client/src/client.rs:2414]: `run_catchup()` is triggered every `catchup_step_period`.
if I'm reading the code right, you are sending `StartSync` requests for every shard every invocation of the function. Is that correct? If yes, isn't that too much?

File: chain/client/src/client_actor.rs [hunk 7]
```diff
@@ -1671,6 +1672,29 @@ impl ClientActor {
...
+                        .get_shard_layout(&epoch_id)
+                        .expect("Cannot get shard layout");
+
+                    let sync_hash = state_sync_status.sync_hash;
+                    for &shard_id in &shards_to_sync {
+                        let shard_uid = ShardUId::from_shard_id_and_layout(shard_id, &shard_layout);
+                        match self.client.state_sync_adapter.clone().read() {
+                            Ok(sync_adapter) => sync_adapter.send(
+                                shard_uid,
+                                (SyncMessage::StartSync(SyncShardInfo { shard_uid, sync_hash }))
```
Comment (nikurt) [inline chain/client/src/client_actor.rs:1689]: Same here. Do you start sync every few milliseconds that this `run_sync_step()` function is triggered?

File: chain/client/src/client.rs (hunk 4 above)
Comment (nikurt) [reply chain/client/src/client.rs:2757]: Thanks for clarifying. Please add it as a comment to the code.

File: chain/client/src/client_actor.rs (hunk 7 above)
Comment (VanBarbascu) [reply chain/client/src/client_actor.rs]: That was unnecessarily too much indeed. I changed it to only be called once.

**Review by nikurt — APPROVED** [review]

**Pull request merged**